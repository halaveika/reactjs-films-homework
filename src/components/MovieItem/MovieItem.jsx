import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Card, Button } from 'antd';
import { CaretRightOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';
import MovieInfo from '../MovieInfo';
import WatchVideoModal from '../WatchVideoModal';
import './MovieItem.scss';

export default function MovieItem({
  id, title, genres, vote_average, poster, overview, getDetails,
}) {
  const navigate = useNavigate();
  const [ownState, setState] = useState(
    {
      hovered: false,
      active: false,
    },
  );

  const toggleHover = () => setState({ ...ownState, hovered: !ownState.hovered, active: false });
  const toggleActive = () => setState({ ...ownState, active: true });

  const handleDetailPageNavigation = async () => {
    await getDetails(id);
    navigate('/details');
    window.scrollTo(0, 0);
  };

  return (
    <Card
      className={`movie-item${(ownState.hovered) ? '__hovered' : ''}`}
      onMouseEnter={toggleHover}
      onMouseLeave={toggleHover}
    >
      <div className={`overlay${(ownState.active) ? '__active' : ''}`} onClick={handleDetailPageNavigation} />
      <img
        className={`poster${(ownState.active) ? '__active' : ''}`}
        alt="Poster!"
        src={poster}
      />
      {

      (ownState.hovered)
        ? (!ownState.active)
          ? (
            <>
              <WatchVideoModal id={id}>
                <Button className="watch-btn" type="primary" shape="circle">
                  <CaretRightOutlined
                    className="icon"
                    style={{ fontSize: '250%' }}
                  />
                </Button>
              </WatchVideoModal>
              <Button className="more-btn" onClick={toggleActive}>View Info</Button>
            </>
          )
          : (
            <WatchVideoModal id={id}>
              <Button className="watch-btn__active" type="primary">Watch Now</Button>
            </WatchVideoModal>
          )

        : ''
        }
      {(ownState.active)
        ? (
          <>
            <MovieInfo
              active
              className="movie-info__active"
              title={title}
              genres={genres}
              vote_average={vote_average}
            />
            <span className="overview">{overview}</span>
          </>
        )
        : <MovieInfo active={false} className="movie-info" title={title} genres={genres} vote_average={vote_average} />}
    </Card>

  );
}

MovieItem.propTypes = {
  id: PropTypes.number.isRequired,
  title: PropTypes.string,
  genres: PropTypes.arrayOf(PropTypes.string),
  vote_average: PropTypes.number,
  poster: PropTypes.string,
  overview: PropTypes.string,
  getDetails: PropTypes.func.isRequired,
};
