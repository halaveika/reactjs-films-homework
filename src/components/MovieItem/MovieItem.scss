.overlay {
  position: absolute;
  display: none;
  width: 200px;
  height: 250px;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.7);
  z-index: 2;
  cursor: pointer;
}

.movie-item,
.movie-item__hovered {
  position: relative;
  width: 200px;
  height: 305px;
  margin: 30px;
  padding: 0;

  &__hovered .overlay {
    display: block;
  }

  &__hovered .watch-btn {
    position: absolute;
    width: 50px;
    height: 50px;
    left: 75px;
    top: 50px;
    z-index: 10;

    &::after {
      content: 'Watch Now';
      position: absolute;
      top: 80px;
      left: -20px;
      font-size: 16px;
      color: #fff;
    }

    & .icon {
      position: absolute;
      top: 7px;
      left: 8px;
    }
  }

  &__hovered .more-btn {
    position: absolute;
    left: 55px;
    top: 190px;
    z-index: 10;
    background-color: transparent;
    color: #fff;
  }
}

.poster {
  width: 200px;
  height: 250px;

  &__active {
    width: 200px;
    height: 250px;
  }
}

.overlay__active {
  position: absolute;
  display: block;
  width: 200px;
  height: 300px;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.7);
  z-index: 2;
  cursor: pointer;
}

.overview {
  display: block;
  position: absolute;
  width: 90%;
  height: 150px;
  font-size: 12px;
  top: 57px;
  left: 15px;
  overflow: hidden;
  color: #fff;
  z-index: 10;
}

.watch-btn__active {
  position: absolute;
  left: 15px;
  bottom: 20px;
  background-color: transparent;
  color: #fff;
  z-index: 10;
}
