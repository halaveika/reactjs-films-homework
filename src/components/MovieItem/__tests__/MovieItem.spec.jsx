import { create, act } from 'react-test-renderer';
import React, { useState } from 'react';
import MovieItem from '../MovieItem';

const realUseState = useState;

const mockMovieItemProps = {
  id: '2412412',
  title: 'Fight club',
  genres: ['drama', 'sport', 'comedy'],
  vote_average: 8,
  poster: 'image_path',
  overview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry',
  video: 'youtube-link',
  handleVideo: jest.fn(),
};

describe('test MovieItem component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const component = create(
    <MovieItem
      id={mockMovieItemProps.id}
      title={mockMovieItemProps.title}
      genres={mockMovieItemProps.genres}
      vote_average={mockMovieItemProps.vote_average}
      poster={mockMovieItemProps.poster}
      overview={mockMovieItemProps.overview}
      video={mockMovieItemProps.video}
      handleVideo={mockMovieItemProps.handleVideo}
    />,
  );

  it('should render MovieItem component', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggleHover should change state', () => {
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => realUseState({
        hovered: true,
        active: false,
      }));

    const component = create(
      <MovieItem
        id={mockMovieItemProps.id}
        title={mockMovieItemProps.title}
        genres={mockMovieItemProps.genres}
        vote_average={mockMovieItemProps.vote_average}
        poster={mockMovieItemProps.poster}
        overview={mockMovieItemProps.overview}
        video={mockMovieItemProps.video}
        handleVideo={mockMovieItemProps.handleVideo}
      />,
    );

    const { root } = component;
    console.dir(root);
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((initialState) => [initialState, setState]);
    act(() => { root.findAllByType('button')[0].props.onClick(); });
    expect(setState.mock.calls.length).toBe(1);
    // expect(setState.mock.calls[0][0]).toStrictEqual({hovered: true, active: false });
  });

  // it('handleVisibleChange should handle button click', () => {
  //   const mockCallback = jest.spyOn(NavBar.prototype, 'handleVisibleChange').mockImplementation();

  //   const { root } = create(
  //     <NavBar
  //       overview={mockData.overview}
  //     />,
  //   );
  //   expect(root.instance.state.visible).toBe(false);
  //   act(() => { root.findAllByType('button')[1].props.onClick(); });
  //   expect(mockCallback.mock.calls.length).toBe(1);
  //   expect(mockCallback.mock.calls[0][0]).toBe(true);
  // });
});
