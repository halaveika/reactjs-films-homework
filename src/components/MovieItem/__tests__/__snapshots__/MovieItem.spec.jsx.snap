// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MovieItem component should render MovieItem component 1`] = `
<div
  className="ant-card ant-card-bordered movie-item"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="overlay"
    />
    <img
      alt="Poster!"
      className="poster"
      src="https://image.tmdb.org/t/p/w500/image_path"
    />
    
    <div
      className="movie-item-container"
    >
      <span
        className="movie-item_title"
      >
        Fight club
      </span>
      <span
        className="movie-item_vote-average"
      >
        8
      </span>
      <span
        className="movie-item_genre"
      >
        drama, sport, comedy
      </span>
    </div>
  </div>
</div>
`;
